File: server/src/db/repositories/tags.py
"""TagRepo + TagGroupRepo — async Repositories over tag-related tables."""

from __future__ import annotations

from typing import TYPE_CHECKING

from db.asyncbridge import in_thread
from db.entities import Tag, TagGroup
from db.helpers import fetch_all_as, fetch_all_dicts, fetch_one_as, sql_placeholders
from services.tag_i18n import translate_tag

if TYPE_CHECKING:
    import sqlite3


class TagRepo:
    def __init__(self, cur: sqlite3.Cursor) -> None:
        self.cur = cur

    @in_thread
    def get(self, name: str) -> Tag | None:
        self.cur.execute(
            "SELECT name, group_id, created_at, updated_at FROM tags WHERE name = ?",
            [name],
        )
        return fetch_one_as(self.cur, Tag)

    @in_thread
    def list_with_counts(
        self,
        *,
        prev: str | None = None,
        limit: int | None = None,
    ) -> list[dict]:
        """Return [{name, group, count}, ...] for ``TagWithCountPublic``.

        ``count`` reads the trigger-maintained ``tags.post_count`` (migration
        0008) instead of GROUP BY-ing the ~9.4M-row ``post_has_tag`` table
        (~630ms per call). That denormalised count covers *canonical* posts
        only — hidden near-duplicate group members are excluded (migration
        0009) — aligning this listing with the tag-filter facet's semantics.
        """
        sql = (
            "SELECT t.name AS name, t.group_id AS group_id, "
            "tg.id AS g_id, tg.name AS g_name, tg.color AS g_color, "
            "t.post_count AS count "
            "FROM tags t "
            "LEFT JOIN tag_groups tg ON tg.id = t.group_id "
        )
        params: list = []
        if prev:
            sql += "WHERE t.name > ? "
            params.append(prev)
        sql += "ORDER BY t.name "
        if limit:
            sql += "LIMIT ?"
            params.append(limit)
        self.cur.execute(sql, params)
        rows = fetch_all_dicts(self.cur)
        return [
            {
                "name": r["name"],
                "group": ({"id": r["g_id"], "name": r["g_name"], "color": r["g_color"]} if r["g_id"] is not None else None),
                "count": r["count"],
            }
            for r in rows
        ]

    @in_thread
    def create(self, name: str, group_id: int | None) -> Tag:
        self.cur.execute(
            "INSERT INTO tags(name, group_id) VALUES(?, ?) ON CONFLICT(name) DO NOTHING",
            [name, group_id],
        )
        self.cur.execute(
            "SELECT name, group_id, created_at, updated_at FROM tags WHERE name = ?",
            [name],
        )
        tag = fetch_one_as(self.cur, Tag)
        if tag is None:
            msg = f"Tag insert failed for: {name}"
            raise RuntimeError(msg)
        return tag

    @in_thread
    def update_group(self, name: str, group_id: int | None) -> Tag | None:
        self.cur.execute(
            "UPDATE tags SET group_id = ?, updated_at = CURRENT_TIMESTAMP WHERE name = ?",
            [group_id, name],
        )
        self.cur.execute(
            "SELECT name, group_id, created_at, updated_at FROM tags WHERE name = ?",
            [name],
        )
        return fetch_one_as(self.cur, Tag)

    @in_thread
    def delete(self, name: str) -> None:
        # post_has_tag.tag_name has ON DELETE CASCADE FK on tags.name,
        # so deleting the tag row cascades the join rows.
        self.cur.execute("DELETE FROM tags WHERE name = ?", [name])

    @in_thread
    def delete_many(self, names: list[str]) -> None:
        if not names:
            return
        ph = sql_placeholders(names)
        self.cur.execute(f"DELETE FROM tags WHERE name IN ({ph})", names)  # noqa: S608

    # ─── Post ↔ tag association ──────────────────────────────────────
    @in_thread
    def add_tag(self, post_id: int, tag_name: str) -> bool:
        """Return True if inserted, False if already existed.

        Collapsed to two statements (down from select + insert + insert):
        the tags-table insert is idempotent (``ON CONFLICT DO NOTHING``),
        and the link-table insert uses ``RETURNING`` so the conflict path
        skips RETURNING — making rowcount/fetchone the existence signal.
        """
        self.cur.execute(
            "INSERT INTO tags(name) VALUES(?) ON CONFLICT DO NOTHING",
            [tag_name],
        )
        self.cur.execute(
            "INSERT INTO post_has_tag(post_id, tag_name, is_auto) VALUES(?, ?, 0) ON CONFLICT DO NOTHING RETURNING post_id",
            [post_id, tag_name],
        )
        return self.cur.fetchone() is not None

    @in_thread
    def remove_tag(self, post_id: int, tag_name: str) -> bool:
        """Return True if removed, False if didn't exist."""
        self.cur.execute(
            "DELETE FROM post_has_tag WHERE post_id = ? AND tag_name = ? RETURNING post_id",
            [post_id, tag_name],
        )
        return self.cur.fetchone() is not None

    def fetch_tags_by_ids(self, ids: list[int], lang: str = "zh-Hans") -> dict[int, list[dict]]:
        """Batch-fetch tags per post, ordered by canonical group then name.

        Synchronous: called from inside the query layer's ``asyncio.to_thread``
        block. Returns the ``PostHasTagPublic``-shaped dicts the read models use.
        ``lang`` picks the ``translated_name`` table (``en`` yields ``None``).
        """
        if not ids:
            return {}
        placeholders = sql_placeholders(ids)
        self.cur.execute(
            f"""
            SELECT pht.post_id AS post_id,
                   pht.is_auto AS is_auto,
                   t.name AS name,
                   t.created_at AS created_at,
                   t.updated_at AS updated_at,
                   tg.id AS group_id,
                   tg.name AS group_name,
                   tg.color AS group_color
            FROM post_has_tag pht
            JOIN tags t ON t.name = pht.tag_name
            LEFT JOIN tag_groups tg ON tg.id = t.group_id
            WHERE pht.post_id IN ({placeholders})
            ORDER BY pht.post_id,
                CASE COALESCE(tg.name, '')
                    WHEN 'artist'    THEN 0
                    WHEN 'copyright' THEN 1
                    WHEN 'character' THEN 2
                    WHEN 'general'   THEN 3
                    WHEN 'meta'      THEN 4
                    ELSE 5
                END,
                t.name
            """,  # noqa: S608
            ids,
        )
        result: dict[int, list[dict]] = {}
        for r in fetch_all_dicts(self.cur):
            result.setdefault(r["post_id"], []).append(
                {
                    "is_auto": bool(r["is_auto"]),
                    "tag_info": {
                        "name": r["name"],
                        "translated_name": translate_tag(r["name"], lang),
                        "created_at": r["created_at"],
                        "updated_at": r["updated_at"],
                        "group": (
                            {
                                "id": r["group_id"],
                                "name": r["group_name"],
                                "color": r["group_color"],
                            }
                            if r["group_id"] is not None
                            else None
                        ),
                    },
                },
            )
        return result


class TagGroupRepo:
    def __init__(self, cur: sqlite3.Cursor) -> None:
        self.cur = cur

    @in_thread
    def get(self, group_id: int) -> TagGroup | None:
        self.cur.execute(
            "SELECT id, name, parent_id, color, created_at, updated_at FROM tag_groups WHERE id = ?",
            [group_id],
        )
        return fetch_one_as(self.cur, TagGroup)

    @in_thread
    def list_all(self) -> list[TagGroup]:
        self.cur.execute(
            "SELECT id, name, parent_id, color, created_at, updated_at FROM tag_groups ORDER BY id",
        )
        return fetch_all_as(self.cur, TagGroup)

    @in_thread
    def ensure(self, name: str, color: str = "#000000") -> TagGroup:
        self.cur.execute(
            "INSERT INTO tag_groups(name, color) VALUES (?, ?) ON CONFLICT(name) DO NOTHING",
            [name, color],
        )
        self.cur.execute(
            "SELECT id, name, parent_id, color, created_at, updated_at FROM tag_groups WHERE name = ?",
            [name],
        )
        tg = fetch_one_as(self.cur, TagGroup)
        if tg is None:
            msg = f"TagGroup upsert failed for: {name}"
            raise RuntimeError(msg)
        return tg

File: packages/db/src/repositories/tags.ts
/**
 * tags / tag_groups 的读取 —— 对应 Python 侧 `db/repositories/tags.py`。
 */
import { placeholders } from '../sql.js'
import type BetterSqlite3 from 'better-sqlite3'

export interface TagGroupRow {
  id: number
  name: string
  color: string
}

export interface TagWithCount {
  name: string
  group: TagGroupRow | null
  count: number
}

/**
 * `[{name, group, count}, ...]`，按 tag 名游标分页。
 *
 * `count` 读的是触发器维护的 `tags.post_count`（migration 0008），而不是对 940 万行的
 * `post_has_tag` 做 GROUP BY（那样每次约 630 ms）。这个反规范化计数**只覆盖 canonical
 * post** —— 近重复分组里被隐藏的成员不计（migration 0009），与 tag 过滤 facet 的语义一致。
 */
export function listTagsWithCounts(
  sqlite: BetterSqlite3.Database,
  { prev, limit }: { prev?: string | null, limit?: number | null } = {},
): TagWithCount[] {
  let sql
    = 'SELECT t.name AS name, t.group_id AS group_id, '
      + 'tg.id AS g_id, tg.name AS g_name, tg.color AS g_color, '
      + 't.post_count AS count '
      + 'FROM tags t '
      + 'LEFT JOIN tag_groups tg ON tg.id = t.group_id '
  const params: unknown[] = []
  if (prev) {
    sql += 'WHERE t.name > ? '
    params.push(prev)
  }
  sql += 'ORDER BY t.name '
  if (limit) {
    sql += 'LIMIT ?'
    params.push(limit)
  }

  const rows = sqlite
    .prepare<unknown[], {
      name: string
      group_id: number | null
      g_id: number | null
      g_name: string | null
      g_color: string | null
      count: number
    }>(sql)
    .all(...params)

  return rows.map(r => ({
    name: r.name,
    group: r.g_id === null ? null : { id: r.g_id, name: r.g_name!, color: r.g_color! },
    count: r.count,
  }))
}

export function listTagGroups(sqlite: BetterSqlite3.Database): TagGroupRow[] {
  return sqlite
    .prepare<[], TagGroupRow>('SELECT id, name, color FROM tag_groups ORDER BY id')
    .all()
}

/** tag 行（`tags` 表本身）。 */
export interface TagRow {
  name: string
  group_id: number | null
  created_at: string
  updated_at: string
}

export function getTag(sqlite: BetterSqlite3.Database, name: string): TagRow | undefined {
  return sqlite
    .prepare<[string], TagRow>('SELECT name, group_id, created_at, updated_at FROM tags WHERE name = ?')
    .get(name)
}

export function getTagGroup(sqlite: BetterSqlite3.Database, id: number): TagGroupRow | undefined {
  return sqlite
    .prepare<[number], TagGroupRow>('SELECT id, name, color FROM tag_groups WHERE id = ?')
    .get(id)
}

export function createTag(
  sqlite: BetterSqlite3.Database,
  name: string,
  groupId: number | null,
): TagRow {
  sqlite
    .prepare('INSERT INTO tags(name, group_id) VALUES(?, ?) ON CONFLICT(name) DO NOTHING')
    .run(name, groupId)
  const tag = getTag(sqlite, name)
  if (!tag)
    throw new Error(`Tag insert failed for: ${name}`)
  return tag
}

export function updateTagGroup(
  sqlite: BetterSqlite3.Database,
  name: string,
  groupId: number | null,
): TagRow | undefined {
  sqlite
    .prepare('UPDATE tags SET group_id = ?, updated_at = CURRENT_TIMESTAMP WHERE name = ?')
    .run(groupId, name)
  return getTag(sqlite, name)
}

/** `post_has_tag.tag_name` 对 `tags.name` 有 ON DELETE CASCADE，join 行自动跟着删。 */
export function deleteTag(sqlite: BetterSqlite3.Database, name: string): void {
  sqlite.prepare('DELETE FROM tags WHERE name = ?').run(name)
}

export function deleteTags(sqlite: BetterSqlite3.Database, names: string[]): void {
  if (!names.length)
    return
  const ph = placeholders(names.length)
  sqlite.prepare(`DELETE FROM tags WHERE name IN (${ph})`).run(...names)
}

/**
 * 给 post 打 tag，插入成功返回 true，本来就有返回 false。
 *
 * 两条语句而不是三条：tags 表的插入是幂等的（`ON CONFLICT DO NOTHING`），关联表
 * 的插入用 `RETURNING` —— 冲突路径不会 RETURNING，于是"有没有取到行"就是存在信号。
 */
export function addTagToPost(
  sqlite: BetterSqlite3.Database,
  postId: number,
  tagName: string,
): boolean {
  sqlite.prepare('INSERT INTO tags(name) VALUES(?) ON CONFLICT DO NOTHING').run(tagName)
  const row = sqlite
    .prepare(
      'INSERT INTO post_has_tag(post_id, tag_name, is_auto) VALUES(?, ?, 0) ON CONFLICT DO NOTHING RETURNING post_id',
    )
    .get(postId, tagName)
  return row !== undefined
}

/** 摘掉 post 上的 tag，删掉了返回 true，本来就没有返回 false。 */
export function removeTagFromPost(
  sqlite: BetterSqlite3.Database,
  postId: number,
  tagName: string,
): boolean {
  const row = sqlite
    .prepare('DELETE FROM post_has_tag WHERE post_id = ? AND tag_name = ? RETURNING post_id')
    .get(postId, tagName)
  return row !== undefined
}
